### ISSUE TITLE
Changing Payment Gateway on Checkout Page doesn't work after trying to use PayPal Gateway

### ISSUE THREAD

**Issue by napoleaowingert** [easydigitaldownloads/easy-digital-downloads#8883]
### Expected behavior
I am filling in my data to proceed with my purchase by using the PayPal payment gateway on checkout page. When I am done and everything is properly filled out, I try to use the PayPal purchase button. It opens a new window that loads the PayPal website. But now I close this window for whatever reason - e.g. I just remember that I can't use PayPal for some reason. So I try now to choose another payment gateway.

When I try to change the payment gateway on checkout page, I expect it to work so I can see the new payment gateway form to proceed with my purchase.

### Actual behavior
But, when I try to change the payment gateway on checkout page, it doesn't display the new payment gateway form. Instead, it doesn't load anything. It is just blank. Also, if I try to go back to the first payment gateway, I get the same problem: no form is displayed.

That can be really frustrating for many customers and they might leave the website without making any purchase as the process doesn't work well.

### Steps to reproduce the behavior
1) Use your website without a logged in user.
2) Try to add some product to your cart.
3) Go to the checkout page.
4) Choose the PayPal payment gateway in case it is not your default.
5) Fill in all the required data correctly.
6) Try to use the PayPal purchase button.
7) Wait for the new window to open and load.
8) Close the PayPal window.
9) Try to change the payment gateway.
10) See your form fade away without any other form being displayed.

### Information (if a specific version is affected):
PHP Version: 8.0.9

EDD Version (or branch): master / 2.11.1

WordPress Version: 5.8

Any other relevant information:
If I am correct, my debug shows that the nonce verification fails - edds_prb_load_gateway function located at easy-digital-downloads/includes/gateways/stripe/includes/payment-methods/payment-request/checkout.php.

It seems that, if the data collected on checkout page is correct and the new PayPal window is open and loaded, a new user is created for the customer - because currently you are not logged in. I am also using Easy Digital Downloads - Auto Register.

There are nonces on checkout page for each payment gateway, generated on page load - at this point you don't have a user.

Later on it creates a new user for you and now you are logged in, but the page is not refreshed - it means that we might have nonces that will not work correctly anymore. It seems that subsequent requests to change the payment gateway fails because of these nonces, that are not maching anymore (client/server).

This happens because nonces takes into account several pieces of data, including the current user - which has changed, right?

So I guess that is the problem right now.

Refreshing the checkout page, after closing the newly opened PayPal window... would that be the solution for this problem?

Thank you.

**Comment by robincornett** [comment]
Thank you for the detailed report and the work to dig into this issue, @napoleaowingert! I have replicated. I think you are right that the auto registration/login is part of the issue here. The issue doesn't happen for me if the user is logged in or if auto register is not active and no account creation is set to process.

We could reload the checkout screen if the PayPal transaction is cancelled, and that does resolve the issue, but I'm marking this for our lead to decide if that is the best way to handle this scenario.

**Comment by dgoldak** [comment]
@napoleaowingert sent in additional details which can be found here:  https://secure.helpscout.net/conversation/1638913522/194711/

**Comment by ashleyfae** [comment]
Yes the problem is definitely that the nonces won't work anymore after the authentication state has changed.

I'd rather avoid a page reload if we can. Perhaps another option to investigate is just sending a new nonce back in one of the PayPal responses, which we could then update in the DOM. Then when it gets used on gateway switch we'd have a new nonce to work with.

(I can just imagine the page reload being irksome if you close the modal, but then for some reason intend on opening it again. :shrug: )

**Comment by napoleaowingert** [comment]
I agree with that, although I'm not fully aware of how things work there with these PayPal responses.

So, if it is possible to update these nonces somehow, I believe it's a good option taking UX into account.
